### PR TITLE
[Filesystem][smartplaylists] Don't sort grouped items by label if random order specified in xsp

### PR DESCRIPTION
**Pull request by the-black-eagle**
## Description
Allow smartplaylists that contain a group tag to be sorted in random order if that is specified in the xsp.

## Motivation and Context

While looking at improving the music video navigation, I noticed that the random artists for music videos that are shown in the right hand pane of Estuary were not random at all but in alphabetical order.  This list is generated from a simple smart playlist included with the skin.
```
<?xml version="1.0" encoding="UTF-8" standalone="yes" ?>
<smartplaylist type="musicvideos">
    <name></name>
    <match>all</match>
    <group>artists</group>
    <order direction="ascending">random</order>
</smartplaylist>
```

 However, because the current code always sorts grouped items by label, the artists are always shown in alphabetical order.  This PR checks to see if a random order was specified and adjusts the sort accordingly in that case.

## How Has This Been Tested?
Tested by mapping a key to reload the skin and verifying that the artist order is indeed different each time.

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by zag2me** [comment]
Hurray, this error has bugged me for years!!

File: xbmc/filesystem/SmartPlaylistDirectory.cpp [hunk 1]
```diff
@@ -282,9 +282,22 @@ namespace XFILE
...
+                       CSettings::SETTING_FILELISTS_IGNORETHEWHENSORTING)
+                       ? SortAttributeIgnoreArticle
+                       : SortAttributeNone);
+      else
+        items.Sort(SortByRandom, SortOrderAscending,
+                   CServiceBroker::GetSettingsComponent()->GetSettings()->GetBool(
+                       CSettings::SETTING_FILELISTS_IGNORETHEWHENSORTING)
+                       ? SortAttributeIgnoreArticle
+                       : SortAttributeNone);
+    }
```
Comment (howie-f) [inline xbmc/filesystem/SmartPlaylistDirectory.cpp:300]: maybe boil this to:
```c
// sort grouped list by label unless random was specified
if (items.Size() > 1 && !group.empty())
{
  items.Sort(playlist.GetOrder() == SortByRandom ? SortByRandom : SortByLabel,
             SortOrderAscending,
             CServiceBroker::GetSettingsComponent()->GetSettings()->GetBool(
                 CSettings::SETTING_FILELISTS_IGNORETHEWHENSORTING)
                 ? SortAttributeIgnoreArticle
                 : SortAttributeNone);
}
```

**Comment by DaveTBlake** [comment]
Relabelled as "music" and "video" as all smart playlists are impacted by this fix (although "grouping" of music smart playlists is very limited). 

I have to wonder if smart playlists _with grouping_ were designed to have the grouping listed in label order. For example albums group by year you see a list of years in year order, or artists by genre the list of genres appears in alphabetcial order.  

I think that the playlist sort order and limits only apply to the list of items below the grouping that you can navigate down to e.g. albums for a year,  and not the list of groups e.g. years, that is displayed for the smart playlist. After all if other sort orders (that have meaning for albums but not years) are not applied, why apply only random order to years too?

So for the music video example smart playlist grouping by artist and sort order random means show a list of artists by name, and then when you click on an artist show the videos in random order. Random (or any other order)  was never intended to apply to the list of groups.

Also noticed that the sort orders offered on the nav screen when viewing an "artist" type smart playlist is totally wrong. This needs tracing and fixing, but not part of this PR specifically.

**Comment by the-black-eagle** [comment]
> For example albums group by year you see a list of years in year order, or artists by genre the list of genres appears in alphabetcial order.

You remember how a few months ago we were discussing the multiple (re)sorting of items?  This is one case where it turns out to be a good thing!  When the lists you mention end up back in the GUI, they get re-sorted into label order (or year etc) anyway.  The random music artists list gets passed to the skin though which shows it in random order.  Even with this change, if you open the exact same playlist in the GUI it gets sorted by name, unless you manually change that in the sideblade.  That is done by GUIViewStateVideo where random is not an option for music artists. 

The same thing applies to years, genres etc.  They will all be re-sorted upon display in that part of the GUI.  This just ensures that the itemlist generated by the xsp is indeed random when it is passed back to the skin, which _can_ show the random order.

**Comment by DaveTBlake** [comment]
> Even with this change, if you open the exact same playlist in the GUI it gets sorted by name, unless you manually change that in the sideblade.

>The same thing applies to years, genres etc. They will all be re-sorted upon display in that part of the GUI.  

Odd,  I tried it (previously had not) and no it doesn't, with music at least. With this change if the smart playlist order is random then the years (for grouped albums) or genres (for grouped artists) get displayed randomly. We are seeing different behaviour, or is it video differing from music?

I don't think the original behaviour (not showing grouping artists randomly) was a bug but by design. It was deliberate, hence the code comment "sort grouped list by label". Lists of grouping items were not intended to be shown randomly. Similarly any limit defined in the xsp is not applied to the list of grouping items. 

If we want to change that then behaviour I think that other sort orders and limits should be considered too (if they have meaning for the grouping field, maybe they don't), and this becomes a feature/improvement rather than a fix.

The problem I see is smart playlist sort order (no grouping) sometimes gets ignored. The sorting and re-sorting in multiple places in the code is not helpful, not only inefficient but it obscures why things come out sorted or not. I have not tested but I would not be surprized if the sorting used when unpacking a smart playlist on  "play" is differnt to that if you navigate to display the list.

**Comment by the-black-eagle** [comment]
Ha, it's been a while for sure !!  Anyway, I think I took @DaveTBlake 's comments on board and the code was modified to only affect music artists of music videos.  To me, that seems specific enough.  Without that change though, here we are, four years later and Estuary is still showing music video artists in alphabetical order, in spite of the label (and smartplaylist) both saying random.

I'll re-test this on top of the current master, but if I see no issues, and no-one has any objections, I'll look to merge this in a week or so.

**Comment by the-black-eagle** [comment]
Sorry @DaveTBlake but I think this is correct!  What we want in this list is a random selection of music video artists, which lead to a (however sorted in GUI) list of videos.  This code produces exactly that result.  I'm happy to discuss the results with you or any other team member, but that said, I'm looking to merge this soon, given no objections that I can't resolve.

**Review by DaveTBlake — APPROVED** [review]
@the-black-eagle this has been hanging around a while, my turn to be sorry for my lack of Kodi activity and attention. The music video behaviour in Estuary bugs you and others and you fix/change it here, all fine by me.

Big picture is the question of sorting grouped items, and unlikely anyone is looking at that any time soon.

**Pull request merged**